00_Base/src/services/credentials.service.ts

```diff
@@ -656,4 +757,41 @@ export class CredentialsService {
     );
     return await clientInformation.save();
   }
+
+  private async getCpoTenantIdForServerRolesForRegistration(
+    serverCredentialsRoleDTOs: CredentialsRoleDTO[],
+  ): Promise<number | undefined> {
+    let cpoTenantId: number | undefined;
+    for (const role of serverCredentialsRoleDTOs) {
+      try {
+        if (role.role !== Role.CPO) {
```

Comment:
I suggest trying to perform all validation in the beginning prior to performing any queries or anything like that. I understand that intuitively we catch errors as we write the code, but in terms of operation execution, your code will benefit from being structured in a way that all validation that is needed in the endpoint handler happens first, prior to execution, which I can understand in some cases may be unavoidable, but I can see that you actually do handle all the error with payload in version right in the beginning. I would suggest creating a `validate` method which than triggers a few more `validateVersion`, `validateRole`, etc. smaller methods. Each method name and definition helps the reader quicker understand what is going on and having smaller methods also helps with testability.

I have an issue with this as well 🤦